feat: Add human friendly comparison periods toggle

rafaeelaudibert opened the pull request:
## Problem

Some customers have seasonal data, where data repeats on specific days of the month/year and they'd like to compare against them. The problem is that if you set "compare against last year" but look at data day to day, then it will be misaligned because the days of the week aren't the same year after year.

## Changes

This toggle allows us to do "human friendly" comparisons, i.e. consider month as 52 weeks and month as 4 weeks. This makes the comparison table slightly nicer to look at, at the cost of not aligning 100% of the time because a month = 30d and a year=365s are slightly more than 4 weeks and 52 weeks, respectively

I've put the configuration inside "Product Analytics" rather than "Web Analytics" because this applies globally - it's a team-based setting.

![image](https://github.com/user-attachments/assets/d7facccb-0bef-4f97-8672-adb98f951a83)

## Does this work well for both Cloud and self-hosted?

Yep

## How did you test this code?

Automated tests + manual UI testing

rafaeelaudibert committed:
feat: Add human friendly comparison periods toggle

This toggle in the backend allows us to do "human friendly" comparisons, i.e. consider month as 52 weeks and month as 4 weeks. This makes the comparison table slightly nicer to look at, at the cost of not aligning 100% of the time because a month = 30d and a year=365s are slightly more than 4 weeks and 52 weeks, respectively

Follow-up commit will add the toggle to the UI

rafaeelaudibert committed:
feat: Add UI to control human friendly comparison periods

You can now toggle that in the UI on `settings/environment-product-analytics#human-friendly-comparison-periods`

rafaeelaudibert commented:
https://github.com/PostHog/posthog/actions/runs/12520844708/job/34927105170?pr=27176 is failing but this is wrong. We can safely add a non-null field with a default on Postgres v11+ - which we are. See https://www.reddit.com/r/programming/comments/bw0vgm/psa_postgresql_11_allows_you_to_create_a_nonnull/

How can we sort this out?

rafaeelaudibert committed:
chore: Silence Django warning on new non-nullable column with default

Postgres 11+ can easily add a new non-nullable column with a default using nothing but an extremely short lock.

https://git.postgresql.org/gitweb/?p=postgresql.git;a=commit;h=16828d5c0273b4fe5f10f42588005f16b415b2d8

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

rafaeelaudibert committed:
fix: Allow field to be nullable

We can't set a default at the DB level in Django 4.2. I really don't understand how this is possible. Rails >>>>>>>> Django.

So, to workaround that, we'll guarantee that we're converting the value to a boolean before sending it to our conversion functions.

rafaeelaudibert committed:
test: Add new value to mock API

rafaeelaudibert committed:
feat: Add new `human_friendly_comparison_toggle` to audit view

rafaeelaudibert commented:
It's not enough for CI to be flaky, GitHub is also failing 
![image](https://github.com/user-attachments/assets/fbc2ca82-30a8-43b3-8db5-6a6d4cb154c8)

rafaeelaudibert committed:
test: Update jest snapshots

rafaeelaudibert committed:
Merge branch 'master' into human-friendly-month-and-year-comparison-periods

robbie-c reviewed (commented):
Period handling code LGTM

See the comment about disabling the warning about adding non-null columns

On the product side - I'm a bit iffy on this. I wonder if we could make the customer happy by changing the default date dropdown from 30 days -> 28 days and adding a 52-week option under 1 year. Wdyt?

rafaeelaudibert commented on posthog/settings/data_stores.py:
```diff
@@ -55,6 +55,13 @@ def postgres_config(host: str) -> dict:
     }
 
 
+# Skip the check for non-nullable fields with default because we're on PG11+
```
Actually, this isn't needed anymore, I'll remove this. Django 4 does not allow us to do this, but I'll be a very happy man once we can upgrade to Django 5.

As for reference on why this is safe, I hope anecdotal evidence is enough: I've done it hundreds (literally) of times before. I checked in Slack and only Michael replied and he was ok with it.

Anyway, this is not needed for now, we can worry about doing this in a safer way once we are on Django 5.

rafaeelaudibert committed:
refactor: Remove unused code

This won't be needed as we can't create a non-null default column in Django 4, so there's no point in disabling the warning if it won't work at all.
